src/Storage/Query/Parser/ContentFieldParser.php

```diff
@@ -0,0 +1,163 @@
+use Bolt\Storage\Query\Conditional\Types;
+use Bolt\Storage\Query\Types\BlockType;
+use Bolt\Storage\Query\Types\DateType;
+use Bolt\Storage\Query\Types\RepeaterType;
+use GraphQL\Type\Definition\IDType;
+use GraphQL\Type\Definition\InputObjectType;
+use GraphQL\Type\Definition\IntType;
+use GraphQL\Type\Definition\ListOfType;
+use GraphQL\Type\Definition\StringType;
+use GraphQL\Type\Definition\Type;
+
+class ContentFieldParser
+{
+    private $config;
+
+    public function __construct(Config $config)
+    {
+        $this->config = $config;
+    }
+
+    public function getParsedContentFields(): array
+    {
+        $contentTypes = $this->config->get('contenttypes');
+        $contentTypesFields = [];
+        /** @var DeepCollection $contentTypeConfiguration */
+        foreach ($contentTypes as $contentType => $contentTypeConfiguration) {
+            $contentTypesFields[$contentType] = $this->parseContentTypeFields($contentType, $contentTypeConfiguration->get('fields'));
+        }
+
+        return $contentTypesFields;
+    }
+
+    public function parseContentTypeFields(string $contentType, DeepCollection $fields): array
+    {
+        $parsedFields = [];
+        foreach ($fields as $fieldName => $fieldConfiguration) {
+            $parsedFields[$fieldName] = $this->getTypeForField($contentType, $fieldConfiguration);
+            $parsedFields['contentType'] = $this->getTypeForField($contentType, new DeepCollection(['type' => 'string']));
+        }
+
+        return $this->prepareConditionalFields($parsedFields);
+    }
+
+    public function getContentFilterFields(string $contentType, bool $isFirst = true): array
+    {
+        $contentFields = $this->getParsedContentFields();
+        $filterFields = [
+            'OR' => [
+                'type' => new ListOfType(Type::nonNull(
+                    new InputObjectType([
+                        'name' => 'OR_filter_'.md5(time().rand(1000,9999)),
+                        'fields' => $isFirst ? array_merge(
+                            $contentFields[$contentType], $this->getContentFilterFields($contentType,false)
+                        ) : $contentFields[$contentType]
+                    ])
+                )),
+            ],
+            'AND' => [
+                'type' => new ListOfType(Type::nonNull(
+                    new InputObjectType([
+                        'name' => 'AND_filter_'.md5(time().rand(1000,9999)),
+                        'fields' => $isFirst ? array_merge(
+                            $contentFields[$contentType], $this->getContentFilterFields($contentType,false)
+                        ) : $contentFields[$contentType]
+                    ])
+                )),
+            ],
+        ];
+
+        return array_merge($filterFields, $contentFields[$contentType]);
+    }
+
+    private function getTypeForField(string $contentType, DeepCollection $fieldConfiguration): Type
+    {
+        switch ($fieldConfiguration['type']) {
```

Comment:
what about user-defined field types?

Reply:
For user-defined field types we should implement some mapper, but it is impossible now, because we cannot implement dynamic DiscriminatorMap for Entity. If new version of doctrine will make it possible, we have to do it, or in the nearest future we have to do by us.